### PR TITLE
Fixes #22: "ComposedObject" Annotation does not add inputSpec to element configuration

### LINKED ISSUE #22: "ComposedObject" Annotation does not add inputSpec to element configuration
I noticed, that my form was not validating correctly. After some digging in code, I found, that although inputSpec is properly set in ElementAnnotationsListener::handleComposedObjectAnnotation, it gets ignored in AnnotationBuilder::configureElement

So here is what I found was causing it, and how I fixed it (although I am not sure, if that is correct):
AnnotationBulder.php:

``` php
    protected function configureElement($annotations, $reflection, $formSpec, $filterSpec)
    {
        // If the element is marked as exclude, return early
        if ($this->checkForExclude($annotations)) {
            return;
        }

        $events = $this->getEventManager();
        $name   = $this->discoverName($annotations, $reflection);

        $elementSpec = new ArrayObject([
            'flags' => [],
            'spec'  => [
                'name' => $name
            ],
        ]);
        $inputSpec = new ArrayObject([
            'name' => $name,
        ]);

        $params = [
            'name'        => $name,
            'elementSpec' => $elementSpec,
            'inputSpec'   => $inputSpec,  // Before the fix, this would not change its vallue
            'formSpec'    => $formSpec,
            'filterSpec'  => $filterSpec,
        ];
        foreach ($annotations as $annotation) {
            $params['annotation'] = $annotation;
// -> This causes ElementAnnotationsListener::handleComposedObjectAnnotation to configure the element
            $events->trigger(__FUNCTION__, $this, $params);
        }
        // Rest of the code is omitted (not relevant)
    }
```

ElementAnnotationsListener.php:

``` php
public function handleComposedObjectAnnotation($e)
    {
        $annotation = $e->getParam('annotation');
        if (!$annotation instanceof ComposedObject) {
            return;
        }

        $class             = $annotation->getComposedObject();
        $annotationManager = $e->getTarget();
        $specification     = $annotationManager->getFormSpecification($class);

        $name        = $e->getParam('name');
        $elementSpec = $e->getParam('elementSpec');

        if ($annotation->isCollection()) {
// Omitted, not relevant
        } else {
            // Compose input filter into parent input filter
            $inputFilter = $specification['input_filter'];
            if (!isset($inputFilter['type'])) {
                $inputFilter['type'] = 'Zend\InputFilter\InputFilter';
            }

// So here is what causes the problem:
// For some reason setting inputSpec like this causes it to be ignored in AnnotationBuilder::configureElement
//            $e->setParam('inputSpec', $inputFilter);
// This seems to work, now the composed object receives the correct inputSpec
            /** @var ArrayObject $inputSpec */
            $inputSpec = $e->getParam('inputSpec');
            $inputSpec->exchangeArray($inputFilter);

            unset($specification['input_filter']);
// Rest of the code is omitted, not relevant
        }
    }
```

I don't believe that this is expected behavior, because every other aspect of the element configuration can be set in ElementAnnotationsListener::handleComposedObjectAnnotation. 

---

Originally posted by @prendit at https://github.com/zendframework/zend-form/issues/125

### PR DESCRIPTION
**Pull request by riul88**
Signed-off-by: Raul Robledo <riul88@gmail.com>

|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | yes
| BC Break      | yes/no
| New Feature   | no
| RFC           | yes/no
| QA            | yes/no

### Description
Fix suggested by @weierophinney 
Issue reproducible by unit test testAllowsComposingChildEntities but the test case didn't validate the filters or validators from the child entity, the unit test was updated as part of this PR to perform the missing assertions to ensure the element validators and filters are configured

The changes to unit test testAllowsComposingMultipleChildEntities which is used when ComposedObject is marked as isCollection are only preventive not impacted by the code change in the PR

### CODE REVIEW

File: test/Annotation/AbstractBuilderTestCase.php
Comment (Ocramius) [inline test/Annotation/AbstractBuilderTestCase.php:245]: Can we split it out to a smaller test highlighting just what is being fixed? :thinking:

**Review by Ocramius — CHANGES_REQUESTED** [review]

File: test/Annotation/AbstractBuilderTestCase.php
Comment (Slamdunk) [reply test/Annotation/AbstractBuilderTestCase.php:245]: Without the fix, `username` and `password` result indeed with no validator at all, wrongfully.

File: test/Annotation/AbstractBuilderTestCase.php
Comment (Slamdunk) [inline test/Annotation/AbstractBuilderTestCase.php:302]: This test passes even without the fix though, but I like it has been added.

**Review by Slamdunk — APPROVED** [review]

**Review by weierophinney — APPROVED** [review]
:ship:

@Slamdunk Since you're official maintainer for this component now, I'll let you merge.

**Comment by Slamdunk** [comment]
Thank you @riul88

**Pull request merged**
